### PR TITLE
Implement the readiness check of the proxy server

### PR DESCRIPTION
**Pull request by caesarxuchao**
It only returns ready if the proxy server has connections to the proxy agents.

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -172,13 +179,15 @@ func (s *ProxyServer) getFrontend(agentID string, connID int64) (*ProxyClientCon
...
 		frontends:                  make(map[string](map[int64]*ProxyClientConnection)),
 		PendingDial:                NewPendingDialManager(),
 		serverID:                   serverID,
 		serverCount:                serverCount,
-		BackendManager:             NewDefaultBackendManager(),
+		BackendManager:             bm,
 		AgentAuthenticationOptions: agentAuthenticationOptions,
+		Readiness:                  bm,
 	}
 	return p
```
Comment (zanetworker) [inline pkg/server/server.go:192]: Maybe simply return without the extra variable? 
```go 
	return  &ProxyServer{
		frontends:                  make(map[string](map[int64]*ProxyClientConnection)),
		PendingDial:                NewPendingDialManager(),
		serverID:                   serverID,
		serverCount:                serverCount,
		BackendManager:             NewDefaultBackendManager(),
		BackendManager:             bm,
		AgentAuthenticationOptions: agentAuthenticationOptions,
		Readiness:                  bm,
	}
```

Comment (caesarxuchao) [reply pkg/server/server.go:192]: Done.

**Comment by caesarxuchao** [comment]
> A change is also needed in the health server to call this as part of the readiness check.

I disagree. I think we should distinguish the healthz from the readiness check. Healthz just means the process is still running.

> I also feel that we don't really need the extra Readiness variable (could just use something like BackendManager.Ready(), but don't have a strong preference.

I don't want to mix the interfaces, so I added another interface and variable.

**Comment by caesarxuchao** [comment]
cc @anfernee

File: pkg/server/server.go [hunk 2]
```diff
@@ -175,15 +182,16 @@ func (s *ProxyServer) getFrontend(agentID string, connID int64) (*ProxyClientCon
...
+	bm := NewDefaultBackendManager()
+	return &ProxyServer{
 		frontends:                  make(map[string](map[int64]*ProxyClientConnection)),
 		PendingDial:                NewPendingDialManager(),
 		serverID:                   serverID,
 		serverCount:                serverCount,
-		BackendManager:             NewDefaultBackendManager(),
+		BackendManager:             bm,
 		AgentAuthenticationOptions: agentAuthenticationOptions,
+		Readiness:                  bm,
```
Comment (cheftako) [inline pkg/server/server.go:193]: Ideally we should have something (ReadinessManager) which aggregates a list of Readiness checks. Any chance we can add the infra now so its easy to additional readiness checks as we find the need?

File: cmd/server/main.go [hunk 3]
```diff
@@ -545,7 +545,7 @@ func (p *Proxy) runHealthServer(o *ProxyRunOptions, server *server.ProxyServer)
 		fmt.Fprintf(w, "ok")
 	})
 	readinessHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		fmt.Fprintf(w, "ok")
```
Comment (cheftako) [inline cmd/server/main.go:548]: Isn't this going to return "true" rather than "ok"?

File: pkg/server/readiness_manager.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+// ReadinessManager supports checking if the proxy server is ready.
+type ReadinessManager interface {
+	// Ready returns if the proxy server is ready.
+	Ready() bool
+}
+
+var _ ReadinessManager = &DefaultBackendManager{}
+
+func (s *DefaultBackendManager) Ready() bool {
```
Comment (cheftako) [inline pkg/server/readiness_manager.go:27]: Would it make more sense for this to return either a message string ("ok" vs "error") or return (bool, string) where the string is a message string? , or even (int, string) where the int is the http response code?

File: pkg/server/server.go (hunk 2 above)
Comment (caesarxuchao) [reply pkg/server/server.go:193]: I did define the ReadinessManager as an interface. Currently I make the backendManager implements the interface. We can compose a dedicated readinessManager with more checks in the future. Is that ok?

**Comment by cheftako** [comment]
> A change is also needed in the health server to call this as part of the readiness check.

+1, It doesn't need to be part of this PR but we re


> > A change is also needed in the health server to call this as part of the readiness check.
> 
> I disagree. I think we should distinguish the healthz from the readiness check. Healthz just means the process is still running.
> 
We should keep liveness (healthz) and readiness checks separate. I think of healthz as the process is more than just running.
It is also in sufficiently healthz state that we don't need Kubelet to restart it. It may be true that if nothing has connected to us in the last ?hour we want to restart, as this may mean something is wrong and that it might be fixed by restarting. In any case we should make sure we have a liveness check as well. (It can be a separate PR)

File: cmd/server/main.go (hunk 3 above)
Comment (cheftako) [reply cmd/server/main.go:548]: Also for non "ok" responses we should be returning an error code via w.WriteHeader().
It needs to be called before we call write. We probably want to make it a 5XX error.

**Comment by Jefftree** [comment]
> > A change is also needed in the health server to call this as part of the readiness check.
> 
> +1, It doesn't need to be part of this PR but we re
> 
> > > A change is also needed in the health server to call this as part of the readiness check.
> > 
> > 
> > I disagree. I think we should distinguish the healthz from the readiness check. Healthz just means the process is still running.
> 
> We should keep liveness (healthz) and readiness checks separate. I think of healthz as the process is more than just running.
> It is also in sufficiently healthz state that we don't need Kubelet to restart it. It may be true that if nothing has connected to us in the last ?hour we want to restart, as this may mean something is wrong and that it might be fixed by restarting. In any case we should make sure we have a liveness check as well. (It can be a separate PR)

That makes sense. Originally I was thinking along the lines of "no agent connection = unhealthy", but I agree that having no agents connected shouldn't force kubelet to restart the pod.

**Comment by caesarxuchao** [comment]
Updated. PTAL. Thanks.

Opened https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/113 to track the discussion about healthz. I agree that "whether we want kubelet to restart the pod" is a good criteria for healthiness.

File: pkg/server/server.go (hunk 2 above)
Comment (cheftako) [reply pkg/server/server.go:193]: Sure

File: pkg/server/server.go [hunk 5]
```diff
@@ -102,6 +102,13 @@ type ProxyServer struct {
 	// BackendManager manages the backends.
 	BackendManager BackendManager
 
+	// Readiness reports if the proxy server is ready, i.e., if the proxy
+	// server has connections to proxy agents (backends). Note that the
+	// proxy server does not check the healthiness of the connections,
+	// though the proxy agents do, so this readiness check might report
+	// ready but there is no healthy connection.
+	Readiness ReadinessManager
```
Comment (cheftako) [inline pkg/server/server.go:110]: Not needed yet but with most K/K liveness/readiness checks we provide a mechanism to disable names checks. I'm guessing we will eventually want this functionality.

**Comment by caesarxuchao** [comment]
Thanks for the reviews!

/hold cancel

**Pull request merged**